[OSX] Application Crash V.0.8.1

isMenuBarInDarkMode() function leads to crash on OSX.

In TrayPosition.kt:82 (saveTrayClickPosition), the code attempts to create or write to tray_position.properties, but this fails because the underlying file system is read-only.

`2025-08-11 16:17:09.535 [AWT-EventQueue-0] [ERROR] caught en error while evaluating tray params: Parent job is Cancelling
kotlinx.coroutines.JobCancellationException: Parent job is Cancelling
Caused by: java.io.FileNotFoundException: tray_position.properties (Read-only file system)
	at java.base/java.io.FileOutputStream.open0(Native Method)
	at java.base/java.io.FileOutputStream.open(Unknown Source)
	at java.base/java.io.FileOutputStream.<init>(Unknown Source)
	at java.base/java.io.FileOutputStream.<init>(Unknown Source)
	at com.kdroid.composetray.utils.TrayPositionKt.saveTrayClickPosition(TrayPosition.kt:82)
	at com.kdroid.composetray.utils.TrayClickTracker.setClickPosition(TrayPosition.kt:43)
	at com.kdroid.composetray.utils.TrayPositionKt.getTrayWindowPosition(TrayPosition.kt:225)
	at MainKt.main$lambda$20$lambda$19(Main.kt:99)
	at androidx.compose.runtime.internal.ComposableLambdaImpl.invoke(ComposableLambda.kt:121)
	at androidx.compose.runtime.internal.ComposableLambdaImpl$invoke$1.invoke(ComposableLambda.kt:122)
	at androidx.compose.runtime.internal.ComposableLambdaImpl$invoke$1.invoke(ComposableLambda.kt:122)
	at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:235)
	at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2838)
	at androidx.compose.runtime.ComposerImpl.skipCurrentGroup(Composer.kt:3158)
	at androidx.compose.runtime.ComposerImpl.doCompose-aFTiNEg(Composer.kt:3706)
	at androidx.compose.runtime.ComposerImpl.recompose-aFTiNEg$runtime(Composer.kt:3648)
	at androidx.compose.runtime.CompositionImpl.recompose(Composition.kt:1002)
	at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:1271)
	at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:142)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2.invokeSuspend$lambda$21(Recomposer.kt:607)
	at androidx.compose.ui.window.YieldFrameClock.withFrameNanos(Application.desktop.kt:268)
	at androidx.compose.ui.window.YieldFrameClock$withFrameNanos$1.invokeSuspend(Application.desktop.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.desktop/java.awt.EventQueue$3.run(Unknown Source)
	at java.desktop/java.awt.EventQueue$3.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.desktop/java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.run(Unknown Source)`

Let me know if you need more information.

--- a/src/commonMain/kotlin/com/kdroid/composetray/utils/AppId.kt
+++ b/src/commonMain/kotlin/com/kdroid/composetray/utils/AppId.kt
@@ -1,0 +1,37 @@
+package com.kdroid.composetray.utils
+
+/**
+ * Provides a unique, stable application identifier to namespace shared resources
+ * (temp files, locks, properties) and avoid conflicts when multiple apps use
+ * this library on the same machine.
+ *
+ * Resolution order (first non-empty wins):
+ * 1) System property "compose.native.tray.appId"
+ * 2) Environment variable "COMPOSE_NATIVE_TRAY_APP_ID"
+ * 3) Main class from system property "sun.java.command" (first token)
+ * 4) Fallback to "ComposeNativeTrayApp"
+ */
+object AppIdProvider {
+    private val cached by lazy { computeAppId() }
+
+    fun appId(): String = cached
+
+    private fun computeAppId(): String {
+        val sunCmd = System.getProperty("sun.java.command")?.trim().orEmpty()
+        println("AppIdProvider: sunCmd: $sunCmd")
+
+        if (sunCmd.isNotEmpty()) {
+            val firstToken = sunCmd.split(" ", limit = 2).firstOrNull().orEmpty()
+            if (firstToken.isNotEmpty()) return sanitize(firstToken)
+        }
+
+        // Fallback
+        return "ComposeNativeTrayApp"
+    }
+
+    private fun sanitize(raw: String): String {
+        // Replace non-alphanumeric/._- with underscore; trim length if excessively long
+        val cleaned = raw.replace(Regex("[^A-Za-z0-9._-]"), "_")
+        return cleaned.take(128).ifEmpty { "ComposeNativeTrayApp" }
+    }
+}

--- a/src/commonMain/kotlin/com/kdroid/composetray/utils/SingleInstanceManager.kt
+++ b/src/commonMain/kotlin/com/kdroid/composetray/utils/SingleInstanceManager.kt
@@ -166,7 +166,7 @@ object SingleInstanceManager {
     }
 
     private fun getAppIdentifier(): String {
-        val callerClassName = Thread.currentThread().stackTrace[3].className
-        return callerClassName.replace(".", "_")
+        // Use unified app ID provider to avoid cross-app conflicts and allow explicit override
+        return AppIdProvider.appId()
     }
 }

--- a/src/commonMain/kotlin/com/kdroid/composetray/utils/TrayPosition.kt
+++ b/src/commonMain/kotlin/com/kdroid/composetray/utils/TrayPosition.kt
@@ -60,38 +60,70 @@ private const val POSITION_KEY = "TrayPosition"
 private const val X_KEY = "TrayX"
 private const val Y_KEY = "TrayY"
 
+// Use a writable tmp/cache directory to avoid read-only filesystem issues (e.g., macOS app bundle working dir)
+private fun trayPropertiesFile(): File {
+    val tmpBase = System.getProperty("java.io.tmpdir") ?: "."
+    val appDir = File(File(tmpBase, "ComposeNativeTray"), AppIdProvider.appId())
+    // Best-effort create directory; if it fails, we will fallback to legacy file access patterns safely
+    runCatching { if (!appDir.exists()) appDir.mkdirs() }.onFailure { /* ignore */ }
+    return File(appDir, PROPERTIES_FILE)
+}
+
+// Legacy properties file in working directory (for backward compatibility)
+private fun legacyPropertiesFile(): File = File(PROPERTIES_FILE)
+
+// Previous tmp location before appId namespacing (for backward-compatible reads only)
+private fun oldTmpPropertiesFile(): File {
+    val tmpBase = System.getProperty("java.io.tmpdir") ?: "."
+    val oldDir = File(tmpBase, "ComposeNativeTray")
+    return File(oldDir, PROPERTIES_FILE)
+}
+
+private fun loadPropertiesFrom(file: File): Properties? {
+    if (!file.exists()) return null
+    return runCatching {
+        Properties().apply { file.inputStream().use { load(it) } }
+    }.getOrNull()
+}
+
+private fun storePropertiesTo(file: File, props: Properties) {
+    // Ensure parent exists if any
+    file.parentFile?.let { runCatching { if (!it.exists()) it.mkdirs() } }
+    runCatching { file.outputStream().use { props.store(it, null) } }
+    // We intentionally swallow exceptions to avoid crashing if the FS is read-only
+}
+
 internal fun saveTrayPosition(position: TrayPosition) {
-    val properties = Properties()
-    val file = File(PROPERTIES_FILE)
-    if (file.exists()) {
-        properties.load(file.inputStream())
-    }
+    val preferredFile = trayPropertiesFile()
+    val properties = loadPropertiesFrom(preferredFile)
+        ?: loadPropertiesFrom(oldTmpPropertiesFile())
+        ?: loadPropertiesFrom(legacyPropertiesFile())
+        ?: Properties()
     properties.setProperty(POSITION_KEY, position.name)
-    file.outputStream().use { properties.store(it, null) }
+    storePropertiesTo(preferredFile, properties)
 }
 
 internal fun saveTrayClickPosition(x: Int, y: Int, position: TrayPosition) {
-    val properties = Properties()
-    val file = File(PROPERTIES_FILE)
-    if (file.exists()) {
-        properties.load(file.inputStream())
-    }
+    val preferredFile = trayPropertiesFile()
+    val properties = loadPropertiesFrom(preferredFile)
+        ?: loadPropertiesFrom(oldTmpPropertiesFile())
+        ?: loadPropertiesFrom(legacyPropertiesFile())
+        ?: Properties()
     properties.setProperty(POSITION_KEY, position.name)
     properties.setProperty(X_KEY, x.toString())
     properties.setProperty(Y_KEY, y.toString())
-    file.outputStream().use { properties.store(it, null) }
+    storePropertiesTo(preferredFile, properties)
 }
 
 internal fun loadTrayClickPosition(): TrayClickPosition? {
-    val file = File(PROPERTIES_FILE)
-    if (!file.exists()) return null
+    // Prefer new location, fallback to old tmp location, then legacy working dir
+    val props = loadPropertiesFrom(trayPropertiesFile())
+        ?: loadPropertiesFrom(oldTmpPropertiesFile())
+        ?: loadPropertiesFrom(legacyPropertiesFile()) ?: return null
 
-    val properties = Properties()
-    properties.load(file.inputStream())
-
-    val positionStr = properties.getProperty(POSITION_KEY) ?: return null
-    val x = properties.getProperty(X_KEY)?.toIntOrNull() ?: return null
-    val y = properties.getProperty(Y_KEY)?.toIntOrNull() ?: return null
+    val positionStr = props.getProperty(POSITION_KEY) ?: return null
+    val x = props.getProperty(X_KEY)?.toIntOrNull() ?: return null
+    val y = props.getProperty(Y_KEY)?.toIntOrNull() ?: return null
 
     return try {
         TrayClickPosition(x, y, TrayPosition.valueOf(positionStr))
@@ -150,11 +182,11 @@ fun getTrayPosition(): TrayPosition {
             }
 
             // Legacy fallback - just position without coordinates
-            val properties = Properties()
-            val file = File(PROPERTIES_FILE)
-            if (file.exists()) {
-                properties.load(file.inputStream())
-                val position = properties.getProperty(POSITION_KEY, null)
+            run {
+                val props = loadPropertiesFrom(trayPropertiesFile())
+                    ?: loadPropertiesFrom(oldTmpPropertiesFile())
+                    ?: loadPropertiesFrom(legacyPropertiesFile())
+                val position = props?.getProperty(POSITION_KEY, null)
                 if (position != null) {
                     return try {
                         TrayPosition.valueOf(position)

